### ISSUE TITLE
RFC: File format loaders

### ISSUE THREAD

**Issue by calebho** [SwatPhonLab/UltraTrace#110]
cc: @keggsmurph21 

Sharing some thoughts on how `ultratrace2.model.files` could be implemented/refactored...

The current approach is to have a class for each kind of file (alignment, image, and sound) containing loader implementations for the accepted file formats. In order to support a new file format for a given file kind, one needs to add a class definition to the respective class. For example, if I want to add support for AAC, I would need to go into `sound.py` and add a class definition under `Sound`.

I propose that we do something a bit different which allows you to physically separate file format implementations from the kind of file they're loading while keeping them logically related using inheritance. We would then maintain a registry which maps MIME type/extension to a loader implementation. The hierarchy would look something like:

* `FileLoaderBase[TFileChunk]`
  * `AlignmentFileLoader`
    * `TextGridFileLoader`
    * ...
  * `ImageFileLoader`
    * `DICOMFileLoader`
    * ...
  * `SoundFileLoader`
    * `WAVFileLoader`
    * `FLACFileLoader`
    * ...

The base class is generic in `TFileChunk` because I suppose the end goal is to have an index operation on FileBundle which return the file chunk at that index for each kind of file. This index operation returns a `TFileChunk` which would be something like `AlignmentChunk` for alignment files, `ImageChunk` for image files, and `SoundChunk` for sound files.

We would then have `registry.py` which contains
* a module-level dictionary mapping an extension to a class which knows how to load that file format and 
* an API to interact with this dictionary.

This API might look something like
```python
def register_loader_for_extension(
    extension: str, loader_cls: Type[FileLoaderBase]
) -> None:
    """Register a loader which recognizes a file in the format indicated by the given 
    extension.

    Parameters:
        extension: A file extension, e.g. ".wav", indicating the file format
        loader_cls: A file loader which knows how to load files in the given file format
    """
    ...

def register_loader_for_mime_type(
    mime_type: str, loader_cls: Type[FileLoaderBase]
) -> None:
    """Analogous to `register_loader_for_extension`, but for MIME types."""
    ...

def get_alignment_loader_for_extension(extension: str) -> AlignmentFileLoader:
    ...

def get_sound_loader_for_extension(extension: str) -> SoundFileLoader:
    ...

def get_image_loader_for_extension(extension: str) -> ImageFileLoader:
    ...

# ... Analogous `get_(alignment|sound|image)_loader_for_mime_type` ...
```

The final piece is dynamically instantiating the correct file loader given a file path. One would parse the file path for the extension, then use the `get_` functions to retrieve the appropriate loader or throw an exception if the file format is not supported. Given the appropriate loader, an instance is constructed using the file path.

Implementing a new file format might look something like
```python
# aac.py
from .registry import register_loader_for_extension, register_loader_for_mime_type
from .base import SoundFileLoader

class AACFileLoader(SoundFileLoader):
    def load(self) -> None:
        ...

    def get_chunk(self, idx: int) -> Optional[SoundChunk]:
        ...

register_loader_for_extension(".aac", AACFileLoader)
register_loader_for_mime_type("audio/aac", AACFileLoader)
```

This approach prevents 1k+ line files as the number of file format implementation grows which the current approach cannot avoid. We can now have one file format implementation per file which is much easier to grok IMO. It also opens up the possibility of exposing the registry API as a public API in which users who don't have control over this library can register their own file loader plugins.

**Comment by keggsmurph21** [comment]
This sounds great.  One small thing is that I think we should have
```python
def get_loader_for_extension(path: str) -> FileLoaderBase: ...
def get_loader_for_mime_type(path: str) -> FileLoaderBase: ...
```
instead of specific ones for each loader subcategory, since the consumer will just be iterating through a tree of files, and each particular file should only be loaded by one `Loader`.

**Comment by keggsmurph21** [comment]
or even
```python
def get_loader_for(path: str) -> FileLoaderBase: ...
```

**Comment by keggsmurph21** [comment]
OK, I'm going to start implementing this and see what falls out.  Thanks for the g00d ideaz

**Comment by keggsmurph21** [comment]
@calebho do you think it makes more sense to return the type or the constructed object? i.e.
```python
def get_loader_for(path: str) -> Type[FileLoaderBase]: ...
# OR
def get_loader_for(path: str) -> FileLoaderBase: ...
```

**Comment by keggsmurph21** [comment]
```python
import mimetypes
import os

from collections import defaultdict
from typing import DefaultDict, Sequence, Set, Type

from .base import FileLoaderBase


# global maps
extension_to_loaders_map: DefaultDict[str, Set[Type[FileLoaderBase]]] = defaultdict(set)
mime_type_to_loaders_map: DefaultDict[str, Set[Type[FileLoaderBase]]] = defaultdict(set)


def register_loader_for_extensions_and_mime_types(
    extensions: Sequence[str],
    mime_types: Sequence[str],
    loader_cls: Type[FileLoaderBase],
) -> None:
    """Register a loader which recognizes a file in the format indicated by the given 
    extensions and MIME types.

    Parameters:
        extensions: A set of file extension, e.g. [".wav"], indicating the file format
        mime_types: A set of MIME types, e.g. ["audio/x-wav", "audio/wav"], also indicating the file format
        loader_cls: A file loader which knows how to load files with the given file extensions and MIME types
    """
    pass


def get_loader_for(path: str) -> Sequence[Type[FileLoaderBase]]:

    global extension_to_loaders_map
    global mime_type_to_loaders_map

    _, extension = os.path.splitext(path.lower())
    mime_type, _ = mimetypes.guess_type(path)

    # NB: Even if `mime_type` is `None`, the `defaultdict` will give us
    #     an empty set, so this won't break.
    loader_clses_by_extension = extension_to_loaders_map[extension]
    loader_clses_by_mime_type = mime_type_to_loaders_map[mime_type]

    # NB: Use set-intersection (could potentially use set-union instead).
    loader_clses = loader_clses_by_extension & loader_clses_by_mime_type

    return loader_clses
```
We may still need to do resolution here, since `loader_clses` could potentially have `len` >1

**Comment by calebho** [comment]
> do you think it makes more sense to return the type or the constructed object?

Yeah I meant to return `Type[...]` originally.

> We may still need to do resolution here, since loader_clses could potentially have len >1

What's the use case for having multiple loaders for a single extension?

**Comment by keggsmurph21** [comment]
> > We may still need to do resolution here, since loader_clses could potentially have len >1
> 
> What's the use case for having multiple loaders for a single extension?

What if someone wants to add their own custom `WAVLoader`? I'm not sure if we should support that, but it's worth consideration.

**Comment by keggsmurph21** [comment]
Also, with https://github.com/SwatPhonLab/UltraTrace/commit/cf17cff973426b050998db3254a24c29a9eda9e2 it turns out we don't want/need generics.  The abstract `ImageSetFileLoader` can just have an abstract `::get_frame(i: int) -> PIL.Image.Image` method.  We don't really need similar functionality from `AlignmentFileLoader` or `SoundFileLoader`.

Tests still need to be written for this stuff of course ...

**Comment by jonorthwash** [comment]
I think @mr-martian is actively working on file reader stuff, so it might be good to coordinate so you don't step on each others' feet.

**Comment by keggsmurph21** [comment]
I'm thinking we'll incorporate that functionality from "upstream", and just
adapt it to the new API.

**Comment by jonorthwash** [comment]
Okay, you guys working in a branch?

**Comment by keggsmurph21** [comment]
Well, we've been committing mostly to the `ultratrace2` directory.  I am currently working in the `manage-multiple-projects` branch (#107), but am hoping to merge that in to `master` soon.

**Comment by mr-martian** [comment]
Two complications that might be relevant:
- There are 3 distinct ways of loading `.dicom` files (reading from pixel data, reading pngs extracted from pixel data, and reading from the private use area)
- Loading `.ult` files actually requires the loader to know about two different files: `.ult` and a metadata file for which I have the extension get set to `US.txt`

**Comment by kmurphy4** [comment]
Do we generate that metadata file ourselves?

**Comment by mr-martian** [comment]
No, it is generated by the recording device.

**Comment by kmurphy4** [comment]
Ok, so we should require that file to also be present in order to decide `.ult` files, right?

**Comment by mr-martian** [comment]
Yes. `xyz.ult` comes with `xyzUS.txt` which is a plain text file containing the image dimensions and a few other things.

**Comment by kmurphy4** [comment]
Thanks! Will keep this in mind

**Comment by calebho** [comment]
> What if someone wants to add their own custom WAVLoader? I'm not sure if we should support that, but it's worth consideration.

Then they would overwrite the default using the registry API. It doesn't make much sense to me to have more than one loader at once for one file format.

**Comment by keggsmurph21** [comment]
> > What if someone wants to add their own custom WAVLoader? I'm not sure if we should support that, but it's worth consideration.
> 
> Then they would overwrite the default using the registry API. It doesn't make much sense to me to have more than one loader at once for one file format.

That's probably a sensible behavior.  I responded to this more fully [here](https://github.com/SwatPhonLab/UltraTrace/pull/109#discussion_r360195178).
